vm/src/builtins/memory.rs

```diff
@@ -54,8 +54,7 @@ type PyMemoryViewRef = PyRef<PyMemoryView>;
 #[pyimpl(with(Hashable, Comparable, AsBuffer))]
 impl PyMemoryView {
     fn parse_format(format: &str, vm: &VirtualMachine) -> PyResult<FormatSpec> {
-        FormatSpec::parse(format)
-            .map_err(|msg| vm.new_exception_msg(vm.ctx.types.memoryview_type.clone(), msg))
+        FormatSpec::parse(format).map_err(|msg| vm.new_struct_error(msg))
```

Comment:
@youknowone the `StructError` is used here. I added this because I ran the UTs before adding any patches and got "all the exceptions thrown must be derived from 'BaseException` (I ccouldn't remember the exactly error info).
I tested it on CPython:
```
>>> import array
>>> a = array.array('u', 'abc')
>>> x = memoryview(a)
>>> x.format
'w'
>>> x[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NotImplementedError: memoryview: format w not supported
```
The behaviour is very different from RustPython.

Before I added `FormatType::WideChar = b'u'`, it just threw an exception when creating the `memoryview`.

Reply:
The original code looks really weird but it also doesn't look like struct error. If you are looking for `NotImplementedError`, it is `vm.new_not_implemented_error`

Reply:
I found where the format `'u'` come from:
https://github.com/python/cpython/blob/3240bc62f4e0afa09964f3afc845697f0a0806b9/Modules/arraymodule.c#L2595-L2599

From the commit message,
> Close #13072: Restore code before the PEP 393 for the array module
> 
> 'u' format of the array module uses again Py_UNICODE type for backward compatibility with Python 3.2.
> ...

I found this issue: https://bugs.python.org/issue13072

Should we support this compatibility? What if the macro `Py_UNICODE_WIDE` is not defined?

My opinion is to modify the type of `PyMemoryView::format_spec` to `PyResult<FormatSpec>`, and then we can throw the error when the `memoryview` is actually used (pack/unpack), instead of when it is created.

Reply:
The latest CPython's doc says that the `array('u')` type is
> Deprecated since version 3.3, will be removed in version 4.0.

See https://docs.python.org/3.11/library/array.html.

I think CPython maintains `array('u')` only for compatibility consideration, and they might not change its behavior because early or late it will be removed.

There are two possible solutions for us:
1. leave the memory view created from `array('u')` unaccepted and throws `NotImplementedError`, but it is not compatible with CPython,
2. implement the memory view format `'u'` ourselves, and let it exposes `array('u')` as a `[wchar_t]` slice.

Which one do you prefer?

Reply:
I vote for 2. And I think this PR is almost doing it :D

Reply:
Yes it is, because I didn't know that CPython does not support formatting of `array('u')`.

Now I have fixed the problems you mentioned before.